ignore/ignore.go

```diff
@@ -0,0 +1,73 @@
+// package ignore provides helpers to support ignore-files similar to .gitignore
+package ignore
+
+import (
+	"bufio"
+	"io"
+	"strings"
+)
+
+var (
+	LineComment = "#"
+	IgnoreFile  = ".sourcegraph/sourcegraphignore"
+)
+
+type IgnoreMatcher struct {
+	IgnoreList []string
+	Strip      int
+}
+
+// StripComponents removes the specified number of leading path
+// elements. Pathnames with fewer elements will return the empty string.
+func StripComponents(path string, count int) string {
+	for i := 0; path != "" && i < count; i++ {
+		i := strings.Index(path, "/")
+		if i < 0 {
+			return ""
+		}
+		path = path[i+1:]
+	}
+	return path
+}
+
+// ParseIgnoreFile parses an ignore-file according to the following rules
+//
+// - each line represents a path relative to the root of the repository
+// - lines starting with # are ignored
+// - empty lines are ignored
+// - if not present, a trailing / is implicit
+func ParseIgnoreFile(r io.Reader) (patterns []string, error error) {
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+		// ignore empty lines
+		if line == "" {
+			continue
+		}
+		// ignore comments
+		if strings.HasPrefix(line, LineComment) {
+			continue
+		}
+		// add trailing "/" to make sure we don't match files that
+		// share a prefix with a directory
+		if !strings.HasSuffix(line, "/") {
+			line += "/"
+		}
+		line = strings.TrimPrefix(line, "/")
+		patterns = append(patterns, line)
+	}
+	return patterns, scanner.Err()
+}
+
+// Match returns true if path has a prefix in common with any item in m.IgnoreList
+func (m *IgnoreMatcher) Match(path string) bool {
```

Comment:
any reason we aren't supporting something more powerful here like the glob work you did?

Reply:
This is the simplest implementation I could think of that fits the requirements. The rules are a subset of  .gitignore and we can extend it if we need to later. The only difference is that we don't support to ignore individual files.

Reply:
@keegancsmith we could use https://github.com/gobwas/glob instead and allow arbitrary glob patterns. It is probably slower than what we have right now but way more powerful. What do you think?

Reply:
edit, I will run a benchmark to quantify my fuzzy statement :)

Reply:
[this benchmark](https://gist.github.com/stefanhengl/6340f78dd211af234014ba53b193d114) times `gobwas.glob` as roughly 38% slower than `strings.hasPrefix`. However, in absolute numbers, they seem both fast enough.

#dirs: 3322
#files: 16610
```
goos: darwin
goarch: amd64
BenchmarkPrefix
BenchmarkPrefix-16    	1000000000	         0.000306 ns/op
BenchmarkGobwas
BenchmarkGobwas-16    	1000000000	         0.000422 ns/op
PASS
```

Reply:
gobwas sounds good. I just remembered that I read the docker sourcecode for dockerignore a long time ago, and it being quite simple. dockerignore and gitignore are basically the same. Maybe we should just use something along those lines and then we also allow exclusion patterns/etc. Not sure if we are getting too complicated, but I can imagine this being useful. I found this package which says it mostly just extracts the docker src code: https://github.com/codeskyblue/dockerignore

Reply:
I think it is best we go with gobwas. It does what we want out of the box, is fast enough, and is used by the community. And we already have it as a dependency in the frontend.